enhancements/operator-dependency-resolution.md

```diff
@@ -0,0 +1,205 @@
+
+An example of a `dependencies.yaml` that specifies Prometheus operator and etcd CRD dependencies:
+
+```
+dependencies:
+  - type: olm.package
+    indexNamespace: io.redhat
+    name: prometheus
+    version: >0.27.0
+  - type: olm.gvk
+    name: etcdclusters.etcd.database.coreos.com
+    group: etcd.database.coreos.com
+    kind: EtcdCluster
+    version: v1beta2
+```
+
+In order to provide a more deterministic dependency resolution, CatalogSource weighting is introduced as a method to order CatalogSource selection when there are multiple CatalogSource providing the same operators and/or APIs. This CatalogSource weighting is set by Cluster Admin. In addition of CatalogSource weighting, the new model will only attempt to resolve dependencies using the default channel. If the specified version of dependency can't be found in default channel, OLM will simply fail to resolve the dependency.
+
+To further enhance the deterministic aspect of dependency resolution, a new field named `indexNamespace` is proposed to enable operator authors to choose a specific source where the operator is provided. This field is essentially a domain/label that can be applied to a specific package or a set of operators. This field combined with operator name will provide a "fully qualified domain name" that is unique for a specific operator. However, the implementation details for this feature is still under development. As a result, this feature is not expected to be included in near-future releases.
+
+#### Provided APIs
+
+The two default types of provided APIs (operator name + version and CRD GVK) will be inferred just by the registry parsing the manifests provided in the bundle. However, we acknowledge that in the future there may be additional APIs provided by operators that are not as easily discoverable.
+
+There are cases when the operators provide/own certain APIs that are important to consider during dependency resolution in order to allow the operator to work properly. This information needs to be declared in the bundle so it can be queried later.
+
+That being said, these additional types will need to be implemented as they become more clear. For now, `dependencies.yaml` will not include any explicit reference to provided APIs.
+
+### Cluster Admin UX
+
+Given the changes to the way OLM will resolve dependencies, it is also necessary to push more detailed data to cluster administrators about what dependencies were installed (or will be installed) and why. To that end, we will continue to use the InstallPlan status to push detailed resolution information to the user. This will include data about what dependencies were installed along with information about where child dependencies came from (essentially, the result of the resolver in detail).
```

Comment:
in CNV we have a kind of meta-operator called hyperconverged cluster operator.
Now we are shipping all the depends operator with a single CSV, the user will only subscribe once to the hyperconverged cluster operator and OLM will deploy all the operators; as soon as the user will create a CR for the hyperconverged operator, that one will create CRs for all the dependent operators.
Because the user is not supposed to directly instantiate dependent operators, we are now hiding them CRDs with `operators.operatorframework.io/internal-objects` annotation.
Can we do something like that (depend on something that is not really that user visible) using this new dependency mechanism?

Than we also need to have CNV as a singleton in the cluster and this is true also for the dependent operators.
Is there a way to enforce this at dependency level?